pallets/loans/src/valuation_method.rs

```diff
@@ -0,0 +1,187 @@
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct DiscountedCashFlows<Rate, Balance> {
+	probability_of_default: Rate,
+	loss_given_default: Rate,
+	discount_rate: Rate,
+}
+
+impl<Rate, Balance> DiscountedCashFlows<Rate, Balance>
+where
+	Rate: FixedPointNumber,
+	Balance: FixedPointOperand + BaseArithmetic,
+{
+	pub fn new(
+		probability_of_default: Rate,
+		loss_given_default: Rate,
+		discount_rate: Rate,
+	) -> Self {
+		Self {
+			probability_of_default,
+			discount_rate,
+			loss_given_default,
+		}
+	}
+
+	/// Calculate the present value based on the discounted future cash flows
+	pub fn present_value(
+		&self,
+		debt: Balance,
+		maturity_date: Moment,
+		origination_date: Option<Moment>,
+		now: Moment,
+		interest_rate_per_sec: Rate,
+	) -> Option<Balance> {
+		if debt.is_zero() {
+			return Some(Balance::zero());
+		}
+
+		// If the loan is overdue, there are no future cash flows to discount,
+		// hence we use the outstanding debt as the value.
+		if now > maturity_date {
+			return Some(debt);
+		}
+
+		// Calculate the expected loss over the term of the loan
+		let tel = Rate::saturating_from_rational(
+			maturity_date - origination_date,
+			math::seconds_per_year(),
+		)
+		.checked_mul(&self.probability_of_default)
+		.and_then(|val| val.checked_mul(&self.loss_given_default))
+		.map(|tel| tel.min(One::one()))?;
+		let tel_inv = Rate::one().checked_sub(&tel)?;
+
+		// Calculate the risk-adjusted expected cash flows
+		let acc_rate = checked_pow(interest_rate_per_sec, (maturity_date - now) as usize)?;
+		let ecf = acc_rate.checked_mul_int(debt)?;
+		let ra_ecf = tel_inv.checked_mul_int(ecf)?;
+
+		// Discount the risk-adjust expected cash flows
+		let rate = checked_pow(self.discount_rate, (maturity_date - now) as usize)?;
+		let d = rate.reciprocal()?;
+		d.checked_mul_int(ra_ecf)
+	}
+
+	/// validates the bullet loan parameters
+	pub fn is_valid(&self, now: Moment) -> bool {
+		vec![
+			// discount should always be >= 1
+			self.discount_rate >= One::one(),
+			// maturity date should always be in future where now is at this instant
+			self.maturity_date > now,
+		]
+		.into_iter()
+		.all(|is_positive| is_positive)
+	}
+}
+
+/// The data structure for outstanding debt based valuation
+#[derive(Encode, Decode, Copy, Clone, PartialEq, TypeInfo)]
+#[cfg_attr(any(feature = "std", feature = "runtime-benchmarks"), derive(Debug))]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct OutstandingDebt<Balance> {}
+
+impl<Balance> OutstandingDebt<Balance>
+where
+	Balance: FixedPointOperand + BaseArithmetic,
+{
+	#[allow(dead_code)]
+	pub fn new() -> Self {
+		Self {}
+	}
+
+	/// calculates the present value of the credit line loan
+	/// https://centrifuge.hackmd.io/uJ3AXBUoQCijSIH9He-NxA#Present-value1
+	/// The debt = current outstanding debt * (1 - written off percentage)
+	pub fn present_value(&self, debt: Balance) -> Option<Balance> {
+		Some(debt)
```

Comment:
For `OutstadingDebt`, is this something that should be implemented or is it the current implementation?

Reply:
Yeah the `present_value` for the `OutstandingDebt` valuation method is just the debt, so this implementation is correct. `The debt = current outstanding debt * (1 - written off percentage)` should be happening before this method already and the `debt` param here should already be the output of this.